### PR TITLE
dev/core#1845 Change FK on civicrm_group to delete the associated gro…

### PR DESCRIPTION
**Pull request by seamuslee001**
…up if a saved search is deleted

Overview
----------------------------------------
This picks up the work from @michaelmcandrew on https://lab.civicrm.org/dev/core/-/issues/1845 and @JKingsnorth on https://lab.civicrm.org/dev/core/-/issues/1772

Before
----------------------------------------
FK sets the saved_search_id column to be NULL when the saved search is deleted leaving the group

After
----------------------------------------
The group is also deleted if the linked saved search is deleted

ping @eileenmcnaughton @colemanw @demeritcowboy

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@colemanw is this still appropriate given that search kit uses saved searches for things other than groups - but maybe it doesn't re-use the SAME saved search

**Comment by mattwire** [comment]
@seamuslee001 Needs rebase

**Comment by eileenmcnaughton** [comment]
We agreed that if the saved search is deleted the group should be deleted -but the reverse should only happen if api_entity is populated

**Comment by seamuslee001** [comment]
if api_entity isn't populated

**Comment by seamuslee001** [comment]
@colemanw @eileenmcnaughton I have updated this now to move the upgrade to 5.35.alpha1 and also to update the delete so it only deletes saved searches that are orphaned that have form_values and no api_params set and the FK is on civicrm_group and now changes rather than setting the saved_search_id column to NULL it now deletes the civicrm_group record if civicrm_saved_search is deleted. I believe this matches what we discussed

**Comment by colemanw** [comment]
Test fail unrelated.
@seamuslee001 it appears the upgrade step is actually doing more now than the runtime code.
The upgrade step in this PR will delete all orphaned, non-api SavedSearches, whereas this PR does not include runtime code to do that.  Correct me if I'm wrong but I don't believe that code already exists in core, which could lead to this upgrade step cleaning things up, only to get messy again.

**Comment by seamuslee001** [comment]
@colemanw yes tho I thought we might be tackling that in a separate PR but you are correct on the implications

**Comment by colemanw** [comment]
@seamuslee001 this needs rebase.

File: CRM/Contact/DAO/Group.php [hunk 1]
```diff
@@ -6,7 +6,7 @@
  *
  * Generated from xml/schema/CRM/Contact/Group.xml
  * DO NOT EDIT.  Generated by CRM_Core_CodeGen
- * (GenCodeChecksum:cb5e8b19ff1f2ff99b556b2e24a67f80)
+ * (GenCodeChecksum:f5ac9e79e86d7fa77408d15da84f3130)
```
Comment (michaelmcandrew) [inline CRM/Contact/DAO/Group.php:9]: I presume that this is incidental?

**Comment by michaelmcandrew** [comment]
@colemanw and @seamuslee001 - I had a go at rebasing this. Not sure about the GenCodeChecksum. I presume that it is not important but let me know if you need me to do anything there.

The checks failed: "No report files were found. Configuration error?". Not sure what I should do about that.

**Comment by demeritcowboy** [comment]
@michaelmcandrew If you click thru on the details and then to console output on the left you'll see

`PHP Fatal error:  Cannot redeclare CRM_Upgrade_Incremental_php_FiveThirtyFive::upgrade_5_35_alpha1() in CRM/Upgrade/Incremental/php/FiveThirtyFive.php on line 77`

**Comment by michaelmcandrew** [comment]
@demeritcowboy thanks! i updated the PR to stick the upgrade step in `CRM_Upgrade_Incremental_php_FiveThirtySix`.

It seems to be building OK now. Fingers crossed this (and not `CRM_Upgrade_Incremental_php_FiveThirtySeven`) is the right place for it. Could you confirm @seamuslee001 ?

**Comment by eileenmcnaughton** [comment]
@seamuslee001 please self-merge this once you get the thumbs up from @demeritcowboy

**Comment by demeritcowboy** [comment]
Actually I'm not even sure why I'm on this. I just responded to help with the gitiness. I'm only partly in tune with what this is about and don't know what was discussed.

**Comment by eileenmcnaughton** [comment]
@demeritcowboy but your comments sounded so knowledgeable!

**Comment by demeritcowboy** [comment]
Fake it till you make it.

Regarding the PR as hinted at earlier in the thread this alone doesn't solve the original problem. Deleting a smart group still leaves the saved search in the system and if it has a custom field which was disabled you still get the status check warning. So should the upgrade be in place before that other part or vice-versa?

And also the upgrade would be in 5.37 now.

But the reverse part is working - if you delete a saved search the group that uses it is deleted.

**Comment by eileenmcnaughton** [comment]
OK yes - this correctly fixes " if you delete a saved search the group that uses it is deleted." and could be merged once the upgrade is (yes @michaelmcandrew ) under 5.37 :-)

However, we need a non-mysql fix to delete the saved search when a group is deleted as this should ONLY happen if the saved search has NULL in api_params

**Comment by seamuslee001** [comment]
thanks @michaelmcandrew and @demeritcowboy for pushing this one along, 

I believe we have now got all the bases covered so this PR now does the following

1. Clears out any stale saved searches i.e. those not linked to a group where there is no api params
2. Implements via the discard function the cleaning out when a smart group is deleted
3. Alters the FK between civicrm_group and civicrm_saved_search so that if a civicrm_saved search the civicrm_group is also deleted.

The upgrade step has also been moved into the 5.37 part as well

**Comment by michaelmcandrew** [comment]
@seamuslee001 - ok grand. Let me know if there is anything else that I can do here to help get it merged in.

File: CRM/Contact/BAO/Group.php [hunk 2]
```diff
@@ -94,6 +96,18 @@ public static function discard($id) {
     $query = "DELETE FROM civicrm_acl_entity_role where entity_table = 'civicrm_group' AND entity_id = %1";
     CRM_Core_DAO::executeQuery($query, $params);
 
+    //check whether this group contains  any saved searches and check if that saved search is appropriate to delete.
+    $groupDetails = Group::get(FALSE)->addWhere('id', '=', $id)->execute();
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Group.php:100]: @seamuslee001 no need to change this but I generally add ->first() at the end & then don't need to refer with [0] in the next rows

**Comment by eileenmcnaughton** [comment]
@michaelmcandrew funny you should ask that - because all this needs now if for someone to do a final test to check it works in the way @seamuslee001 describes above.....

File: CRM/Upgrade/Incremental/php/FiveThirtySeven.php [hunk 3]
```diff
@@ -69,4 +79,16 @@ public function setPostUpgradeMessage(&$postUpgradeMessage, $rev) {
...
 
+  /**
+   * @param \CRM_Queue_TaskContext $ctx
+   *
+   * @return bool
+   */
+  public static function alterSavedSearchFK(CRM_Queue_TaskContext $ctx) {
+    CRM_Core_BAO_SchemaHandler::safeRemoveFK('civicrm_group', 'FK_civicrm_group_saved_search_id');
+    CRM_Core_DAO::executeQuery('DELETE civicrm_saved_search FROM civicrm_saved_search LEFT JOIN civicrm_group ON civicrm_saved_search.id = civicrm_group.saved_search_id WHERE civicrm_group.id IS NULL AND form_values IS NOT NULL and api_params IS NULL');
+    CRM_Core_DAO::executeQuery('ALTER TABLE civicrm_group ADD CONSTRAINT `FK_civicrm_group_saved_search_id` FOREIGN KEY (`saved_search_id`) REFERENCES `civicrm_saved_search`(`id`) ON DELETE CASCADE', [], TRUE, NULL, FALSE, FALSE);
```
Comment (demeritcowboy) [inline CRM/Upgrade/Incremental/php/FiveThirtySeven.php:90]: Up above in the group delete it also checks api_entity. Here it only checks api_params. Is it possible for one to be empty and not the other?

**Comment by demeritcowboy** [comment]
Other than the upgrade script comment above I tested deleting both ways and it works as described.

**Comment by eileenmcnaughton** [comment]
Merging based on @demeritcowboy review - oh just read the upgrade comment - I think it's safe

**Pull request merged**